src/metabase/util/ssh.clj

```diff
@@ -0,0 +1,95 @@
+(ns metabase.util.ssh
+  (:require [clojure.tools.logging :as log]
+            [metabase.util :as u])
+  (:import com.jcraft.jsch.JSch))
+
+(def default-ssh-timeout 30000)
+
+(defn start-ssh-tunnel
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass host port]}]
+  (let [connection (doto (.getSession (new com.jcraft.jsch.JSch) tunnel-user tunnel-host tunnel-port)
+                     (.setPassword tunnel-pass)
+                     (.setConfig "StrictHostKeyChecking" "no")
+                     (.connect default-ssh-timeout)
+                     (.setPortForwardingL 0 host port))
+        input-port (some-> (.getPortForwardingL connection)
+                           first
+                           (clojure.string/split #":")
+                           first
+                           (Integer.))]
+    (assert (number? input-port))
+    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
+    [connection input-port]))
+
+(def ssh-tunnel-preferences
+  [{:name         "tunnel-enabled"
+    :display-name "Use SSH tunnel"
+    :placeholder  "Enable this ssh tunnel?"
+    :type         :boolean
+    :default      false}
+   {:name         "tunnel-host"
+    :display-name "SSH tunnel host"
+    :placeholder  "What hostname do you use to connect to the SSH tunnel?"}
+   {:name         "tunnel-port"
+    :display-name "SSH tunnel port"
+    :type         :integer
+    :default      22}
+   {:name         "tunnel-user"
+    :display-name "SSH tunnel username"
+    :placeholder  "What username do you use to login to the SSH tunnel?"}
+   {:name         "tunnel-pass"
+    :display-name "SSH tunnel password"
+    :type         :password
+    :placeholder  "******"}
+   #_{:name         "tunnel-private-key"
+    :display-name "SSH private key to connect to the tunnel"
+    :type         :string
+    :placeholder  "Paste the contents of an ssh private key here"}
+   #_{:name         "tunnel-private-key-file-name"
+    :display-name "Path on the Metabase server to a SSH private key file to connect to the tunnel"
+    :type         :string
+    :placeholder  "/home/YOUR-USERNAME/.ssh/id_rsa"}])
+
+(defn with-tunnel-config
+  "Add preferences for ssh tunnels to a drivers :details-fields"
+  [driver-config-map]
+  (assoc driver-config-map :details-fields
+    (constantly (concat ((:details-fields driver-config-map))
+                        ssh-tunnel-preferences))))
```

Comment:
Instead of doing `with-tunnel-config` for the entire driver couldn't we just do

```clojure
;; some driver method implementation map
{:details-fields (merge ssh/ssh-tunnel-preferences
                                    ... ; other things already in the driver
                        )}
```

seems like it would avoid an extra level of indentation everywhere as well as make things a little clearer since you wouldn't have to hop over to this namespace and read this function to understand exactly what's going on